File: sdks/elixir/lib/openrate/http.ex
defmodule Openrate.HTTP do
  @moduledoc """
  A minimal HTTP/1.1 GET over `:gen_tcp`, so this package has no dependency —
  not even `:inets`, which drags in `:ssl` and `:public_key` before it will
  answer a plain-http request.

  Everything openrate's JSON API exposes is a GET against loopback, which is why
  something this small is enough. Point a real HTTP client at
  `Openrate.base_url/0` if you want pooling, retries or TLS.
  """

  @doc """
  `GET base <> path`. Returns `{:ok, status, body}` or `{:error, reason}`.

  The socket is closed on every path, including a failed read — that is what the
  `try/after` is for.
  """
  @spec get(String.t(), String.t(), timeout()) ::
          {:ok, non_neg_integer(), binary()} | {:error, term()}
  def get(base, path, timeout \\ 30_000)

  def get("http://" <> hostport, path, timeout) do
    [host, port] = String.split(hostport, ":", parts: 2)

    case :gen_tcp.connect(
           String.to_charlist(host),
           String.to_integer(port),
           [:binary, active: false, packet: :raw],
           5_000
         ) do
      {:ok, sock} ->
        try do
          request = [
            "GET ",
            path,
            " HTTP/1.1\r\nHost: ",
            hostport,
            "\r\nAccept: application/json\r\nConnection: close\r\n\r\n"
          ]

          with :ok <- :gen_tcp.send(sock, request),
               {:ok, raw} <- read_all(sock, "", timeout) do
            parse(raw)
          end
        after
          :gen_tcp.close(sock)
        end

      {:error, reason} ->
        {:error, reason}
    end
  end

  def get(base, _path, _timeout), do: {:error, {:unsupported_scheme, base}}

  defp read_all(sock, acc, timeout) do
    case :gen_tcp.recv(sock, 0, timeout) do
      {:ok, more} -> read_all(sock, acc <> more, timeout)
      {:error, :closed} -> {:ok, acc}
      {:error, reason} -> {:error, reason}
    end
  end

  # `Connection: close` means the body runs to EOF, so the whole response is in
  # hand by the time this is called and no chunk decoding is needed.
  defp parse(raw) do
    case :binary.split(raw, "\r\n\r\n") do
      [head, body] ->
        ["HTTP/1." <> <<_::binary-size(1)>> <> " " <> <<code::binary-size(3)>> <> _ | headers] =
          String.split(head, "\r\n")

        body =
          if Enum.any?(headers, &(String.downcase(&1) =~ "transfer-encoding: chunked")) do
            dechunk(body, "")
          else
            body
          end

        {:ok, String.to_integer(code), body}

      _ ->
        {:error, :malformed_response}
    end
  end

  defp dechunk(rest, acc) do
    case :binary.split(rest, "\r\n") do
      [size_line, tail] ->
        case Integer.parse(String.trim(size_line), 16) do
          {0, _} ->
            acc

          {size, _} ->
            <<chunk::binary-size(size), "\r\n", tail::binary>> = tail
            dechunk(tail, acc <> chunk)

          :error ->
            acc
        end

      _ ->
        acc
    end
  end
end

File: sdks/rust/src/http.rs
//! A deliberately tiny HTTP/1.1 client for talking to the sidecar.
//!
//! This crate has **no HTTP dependency**, and that is a decision rather than an
//! oversight. openrate's whole API over HTTP is four read-only GETs against
//! `127.0.0.1`. Pulling in `reqwest` for that would drag `hyper`, `tokio` and a
//! TLS stack into every consumer of a crate whose main job is to `dlopen` a
//! shared library.
//!
//! What you are looking at is therefore **not** a general HTTP client and must
//! not be used as one. It speaks to a loopback address, does no TLS, no
//! redirects, no proxies, no chunked-transfer decoding beyond what the sidecar
//! emits, and no connection reuse. If your program already depends on
//! `reqwest`, use `reqwest` — the wire contract is ordinary JSON over HTTP and
//! nothing here is privileged.

use std::io::{BufRead, BufReader, Read, Write};
use std::net::TcpStream;
use std::time::Duration;

/// A failure talking to the sidecar.
#[derive(Debug)]
pub enum HttpError {
    /// The URL was not `http://host:port/path`.
    BadUrl(String),
    /// Connect, read or write failed.
    Io(std::io::Error),
    /// A non-200 response. Carries the code and the body, because the body is
    /// where openrate puts its `{"error": "..."}` object.
    Status(u16, String),
}

impl std::fmt::Display for HttpError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            HttpError::BadUrl(u) => write!(f, "not an http:// url: {u}"),
            HttpError::Io(e) => write!(f, "io: {e}"),
            HttpError::Status(c, b) => write!(f, "HTTP {c}: {b}"),
        }
    }
}

impl std::error::Error for HttpError {}

impl From<std::io::Error> for HttpError {
    fn from(e: std::io::Error) -> Self {
        HttpError::Io(e)
    }
}

fn split_url(url: &str) -> Result<(String, String), HttpError> {
    let rest = url
        .strip_prefix("http://")
        .ok_or_else(|| HttpError::BadUrl(url.to_string()))?;
    match rest.find('/') {
        Some(i) => Ok((rest[..i].to_string(), rest[i..].to_string())),
        None => Ok((rest.to_string(), "/".to_string())),
    }
}

fn connect(hostport: &str, timeout: Duration) -> Result<TcpStream, HttpError> {
    let stream = TcpStream::connect(hostport)?;
    stream.set_read_timeout(Some(timeout))?;
    stream.set_write_timeout(Some(timeout))?;
    Ok(stream)
}

fn send(
    stream: &mut TcpStream,
    method: &str,
    hostport: &str,
    path: &str,
    accept: &str,
    body: Option<&str>,
) -> Result<(), HttpError> {
    let mut req = format!(
        "{method} {path} HTTP/1.1\r\nHost: {hostport}\r\nAccept: {accept}\r\nConnection: close\r\n"
    );
    if let Some(b) = body {
        req.push_str("Content-Type: application/json\r\n");
        req.push_str(&format!("Content-Length: {}\r\n", b.len()));
    }
    req.push_str("\r\n");
    stream.write_all(req.as_bytes())?;
    if let Some(b) = body {
        stream.write_all(b.as_bytes())?;
    }
    stream.flush()?;
    Ok(())
}

/// Reads the status line and headers, leaving the reader positioned at the
/// body. Returns the status code.
fn read_head<R: BufRead>(r: &mut R) -> Result<u16, HttpError> {
    let mut line = String::new();
    r.read_line(&mut line)?;
    let code = line
        .split_whitespace()
        .nth(1)
        .and_then(|c| c.parse::<u16>().ok())
        .ok_or_else(|| HttpError::BadUrl(format!("bad status line: {}", line.trim())))?;
    loop {
        let mut h = String::new();
        if r.read_line(&mut h)? == 0 || h == "\r\n" || h == "\n" {
            break;
        }
    }
    Ok(code)
}

/// GET a URL and return the body, erroring on any non-200.
pub fn get(url: &str, timeout: Duration) -> Result<String, HttpError> {
    request("GET", url, None, timeout)
}

/// POST a JSON body and return the response body, erroring on any non-200.
pub fn post_json(url: &str, body: &str, timeout: Duration) -> Result<String, HttpError> {
    request("POST", url, Some(body), timeout)
}

fn request(
    method: &str,
    url: &str,
    body: Option<&str>,
    timeout: Duration,
) -> Result<String, HttpError> {
    let (hostport, path) = split_url(url)?;
    let mut stream = connect(&hostport, timeout)?;
    send(
        &mut stream,
        method,
        &hostport,
        &path,
        "application/json",
        body,
    )?;

    let mut reader = BufReader::new(&mut stream);
    let code = read_head(&mut reader)?;
    let mut out = String::new();
    reader.read_to_string(&mut out)?;
    if code != 200 {
        return Err(HttpError::Status(code, out.trim().to_string()));
    }
    Ok(out)
}

/// POST a JSON body and stream the `text/event-stream` response, calling
/// `on_chunk` with each `data:` payload and stopping at `[DONE]`.
///
/// Return `true` from `on_chunk` to keep reading and `false` to stop. Returns
/// the number of chunks delivered.
///
/// openrate has no streaming API, so nothing in this crate calls this. It is
/// kept because `http` is shared verbatim with llmux's Rust SDK, where chat
/// streaming is the main event, and because a divergent copy of a hand-written
/// HTTP client is worse than an unused function.
pub fn post_sse<F>(
    url: &str,
    body: &str,
    timeout: Duration,
    mut on_chunk: F,
) -> Result<usize, HttpError>
where
    F: FnMut(&str) -> bool,
{
    let (hostport, path) = split_url(url)?;
    let mut stream = connect(&hostport, timeout)?;
    send(
        &mut stream,
        "POST",
        &hostport,
        &path,
        "text/event-stream",
        Some(body),
    )?;

    let mut reader = BufReader::new(&mut stream);
    let code = read_head(&mut reader)?;
    if code != 200 {
        let mut out = String::new();
        reader.read_to_string(&mut out)?;
        return Err(HttpError::Status(code, out.trim().to_string()));
    }

    let mut n = 0usize;
    let mut line = String::new();
    loop {
        line.clear();
        if reader.read_line(&mut line)? == 0 {
            break;
        }
        let line = line.trim();
        let Some(payload) = line.strip_prefix("data:") else {
            continue;
        };
        let payload = payload.trim();
        if payload == "[DONE]" {
            break;
        }
        n += 1;
        if !on_chunk(payload) {
            break;
        }
    }
    Ok(n)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn split_url_splits_host_and_path() {
        assert_eq!(
            split_url("http://127.0.0.1:8080/v1/models").unwrap(),
            ("127.0.0.1:8080".to_string(), "/v1/models".to_string())
        );
        assert_eq!(
            split_url("http://127.0.0.1:8080").unwrap(),
            ("127.0.0.1:8080".to_string(), "/".to_string())
        );
    }

    #[test]
    fn split_url_rejects_https() {
        // Refusing loudly beats connecting to port 443 and speaking plaintext.
        assert!(matches!(
            split_url("https://example.com/"),
            Err(HttpError::BadUrl(_))
        ));
    }

    #[test]
    fn read_head_parses_status_and_consumes_headers() {
        let raw = b"HTTP/1.1 404 Not Found\r\nContent-Type: text/plain\r\n\r\nbody here";
        let mut r = BufReader::new(&raw[..]);
        assert_eq!(read_head(&mut r).unwrap(), 404);
        let mut rest = String::new();
        r.read_to_string(&mut rest).unwrap();
        assert_eq!(rest, "body here");
    }
}
